miku-server/src/main/scala/org/http4s/miku/MikuBuilder.scala

```diff
@@ -0,0 +1,492 @@
+
+  protected[this] def newRequestHandler(): ChannelInboundHandler =
+    if (enableWebsockets)
+      MikuHandler.websocket[F](httpService, serviceErrorHandler)
+    else
+      MikuHandler.default[F](httpService, serviceErrorHandler)
+
+  /** Return our SSL context.
+    *
+    */
+  private[this] def getContext(): Option[(SSLContext, ClientAuth)] = sslBits.map {
+    case MikuKeyStoreBits(keyStore, keyManagerPassword, protocol, trustStore, clientAuth) =>
+      val ksStream = new FileInputStream(keyStore.path)
+      val ks = KeyStore.getInstance("JKS")
+      ks.load(ksStream, keyStore.password.toCharArray)
+      ksStream.close()
+
+      val tmf = trustStore.map { auth =>
+        val ksStream = new FileInputStream(auth.path)
+
+        val ks = KeyStore.getInstance("JKS")
+        ks.load(ksStream, auth.password.toCharArray)
+        ksStream.close()
+
+        val tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm)
+
+        tmf.init(ks)
+        tmf.getTrustManagers
+      }
+
+      val kmf = KeyManagerFactory.getInstance(
+        Option(Security.getProperty("ssl.KeyManagerFactory.algorithm"))
+          .getOrElse(KeyManagerFactory.getDefaultAlgorithm)
+      )
+
+      kmf.init(ks, keyManagerPassword.toCharArray)
+
+      val context = SSLContext.getInstance(protocol)
+      context.init(kmf.getKeyManagers, tmf.orNull, null)
+
+      (context, clientAuth)
+
+    case MikuSSLContextBits(context, clientAuth) =>
+      (context, clientAuth)
+  }
+
+  /** A stream transformation that registers our channels and
+    * adds the necessary codecs and transport handlers
+    */
+  def channelPipe(
+      eventLoop: MultithreadEventLoopGroup,
+      allChannels: DefaultChannelGroup): Pipe[F, Channel, Unit] = { s =>
+    s.evalMap { connChannel =>
+      F.delay {
+
+        // Setup the channel for explicit reads
+        connChannel
+          .config()
+          .setOption(ChannelOption.AUTO_READ, java.lang.Boolean.FALSE)
+
+        val pipeline = connChannel.pipeline()
+        getContext() match {
+          case Some((ctx, auth)) => //Ignore client authentication for now
+            val engine = ctx.createSSLEngine()
+            engine.setUseClientMode(false)
+            auth match {
+              case NoClientAuth => ()
+              case ClientAuthRequired =>
+                engine.setNeedClientAuth(true)
+              case ClientAuthOptional =>
+                engine.setWantClientAuth(true)
+            }
+            pipeline.addLast("ssl", new SslHandler(engine))
+
+          case None => //Do nothing
+        }
+
+        // Netty HTTP decoders/encoders/etc
+        pipeline.addLast(
+          "decoder",
+          new HttpRequestDecoder(maxInitialLineLength, maxHeaderSize, maxChunkSize))
+        pipeline.addLast("encoder", new HttpResponseEncoder())
+        pipeline.addLast("decompressor", new HttpContentDecompressor())
+
+        idleTimeout match {
```

Comment:
Isn't this non-exhaustive on `Duration.NegativeInf` and `Duration.Undefined`?  (Not that anyone is likely to set those, but...) I usually use `isFinite` instead of pattern matching.  I've not seen this idiom.